### PR TITLE
Adds EventSourcedEntity with EnforcingReplies

### PR DESCRIPTION
**Pull request by octonato**
Still need to write tests.

### CODE REVIEW

**Review by patriknw — COMMENTED** [review]
good, it was an oversight to not have that combination

File: akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/EventSourcedEntity.scala
Comment (renatocaval) [reply akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/EventSourcedEntity.scala]: I can add those on this same PR and some test to prove that things compile at least. Maybe not necessary to write a full functional test for all those cases because we are only doing type massage.

Comment (patriknw) [reply akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/EventSourcedEntity.scala]: Yes, compilation tests would be enough

Comment (renatocaval) [reply akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/EventSourcedEntity.scala]: I added a compilation only test for the enforced replies case.

And `of` variants for it as well.

**Comment by renatocaval** [comment]
@patriknw, I fixed the tests. This is ready for another review. Thanks.

File: akka-cluster-sharding-typed/src/test/java/jdocs/akka/cluster/sharding/typed/HelloWorldPersistentEntityExample.java
Comment (patriknw) [inline akka-cluster-sharding-typed/src/test/java/jdocs/akka/cluster/sharding/typed/HelloWorldPersistentEntityExample.java]: Mention this rename in akka-docs/src/main/paradox/project/migration-guide-2.5.x-2.6.x.md
We will be more strict with API changes from now.

Comment (renatocaval) [reply akka-cluster-sharding-typed/src/test/java/jdocs/akka/cluster/sharding/typed/HelloWorldPersistentEntityExample.java]: ok, I will do it

File: akka-docs/src/main/paradox/project/migration-guide-2.5.x-2.6.x.md
Comment (renatocaval) [inline akka-docs/src/main/paradox/project/migration-guide-2.5.x-2.6.x.md]: @patriknw, added this API change notice. I have looked around to see other examples of API change notices for APIs makes as 'may change', but didn't find any. 

I'm not so sure about the wording. So, any feedback is welcome. 

I also added new methods to the list. Maybe not strictly needed, but looks nice to have all the changes listed, IMO.

**Review by patriknw — APPROVED** [review]
LGTM, thanks

File: akka-docs/src/main/paradox/project/migration-guide-2.5.x-2.6.x.md
Comment (renatocaval) [inline akka-docs/src/main/paradox/project/migration-guide-2.5.x-2.6.x.md]: there is a new commit on master adding an "Akka Typed" section. I'm moved "Akka Typed API changes" to be a sub-header of "Akka Typed"

**Comment by renatocaval** [comment]
Can we have a second sign-off here?

Not that we need it in upcoming 2.5.23, but to avoid potential merge conflicts. If it stales we only increase the chance.

File: akka-cluster-sharding-typed/src/test/java/akka/cluster/sharding/typed/javadsl/ShardingEventSourcedEntityWithEnforcedRepliesCompileOnlyTest.java
Comment (johanandren) [inline akka-cluster-sharding-typed/src/test/java/akka/cluster/sharding/typed/javadsl/ShardingEventSourcedEntityWithEnforcedRepliesCompileOnlyTest.java]: Since it is a compile only test that will never be run you could skip the above and let `ClusterSharding sharding = null`.

File: akka-cluster-sharding-typed/src/test/java/akka/cluster/sharding/typed/javadsl/ShardingEventSourcedEntityWithEnforcedRepliesCompileOnlyTest.java
Comment (johanandren) [inline akka-cluster-sharding-typed/src/test/java/akka/cluster/sharding/typed/javadsl/ShardingEventSourcedEntityWithEnforcedRepliesCompileOnlyTest.java]: Not needed for a compile-only test, I think this will spin up an actor system that is then never used.

**Review by johanandren — APPROVED** [review]
LGTM with a bunch of nitpicks

File: akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/EventSourcedEntity.scala
Comment (renatocaval) [reply akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/javadsl/EventSourcedEntity.scala]: I added FQCN for `Behavior`. `EventSourcedBehavior` is not needed because it's already imported.

**Review by octonato — COMMENTED** [review]
Thanks for the review @johanandren.

I applied the requested changes. Except the usage of `val` vs. `private` in `EventSourcedEntityWithEnforcedReplies` constructor. I'm following what was done for `EventSourcedEntity` which I believe was a conscious decision. If not, we can/should review it.

**Pull request merged**
